design/open/high-level-python-wrapper.md

```diff
@@ -0,0 +1,314 @@
+# lakefs high level Python SDK Wrapper
+
+## Goals
+
+1. Provide a simpler programming interface with less configuration
+1. Behave closer to other related Python SDKs (Pinecone, OpenAI, HuggingFace, ...)
+1. Allow infering identity from environment
+1. Provide better abstractions for common, more complex operations (I/O, imports)
+
+## Non-Goals
+
+1. For now, we explicitly leave out environment administration: setting up server, configuring GC rules, all IAM/ACL related things.
+
+## Authentication
+
+Any operation that calls out to lakeFS will try to authenticate using the following chain:
+
+1. All models receive an optional `client` kwarg with explicit credentials
+1. Otherwise, if `lakefs.init(...)` is called with parameters (`access_key_id`, `jwt_token`, ...) - these will be set on a `lakefs.DefaultClient` object
+1. If `lakefs.init()` is called with no parameters, or `init` is not called:
+    1. use `LAKECTL_ACCESS_KEY_ID` and `LAKECTL_ACCESS_SECRET_KEY` if set
+    1. Otherwise, use `~/.lakectl.yaml` if exists
+    1. Otherwise, try and use IAM role from current machine (using AWS IAM role (will only work with enterprise/cloud))
+1. If init is not called, it will be lazily called on the first use of `DefaultClient`, deferring authentication to the first API call.
+
+## API wrapper interface
+
+the higher lavel SDK will be resource-class based. performing API operations will be
+done by calling the methods on their parent object. Examples:
+
+```python
+import lakefs
+
+repo = lakefs.repository('example')
+branch = repo.branch('main')
+for item in branch.list_objects(prefix='foo/'):
+    if item.path.endswith('.parquet'):
+        print(item.path)
+
+data: bytes = branch.download('datasets/foo/1.parquet')
+branch.upload(data, 'foo/bar/baz.parquet')
+
+# this will work:
+data: bytes = lakefs.repository('example').commit('abc123').download('a/b.txt')
+
+# since commits are immutable, this method will not exist:
+lakefs.repository('example').commit('abc123').upload(data, 'a/b.txt')
+```
+
+## Partial interface definition
+
+### Authentication
+
+```python
+class Client:
+    """
+    Wrapper around lakefs_sdk's client object
+    Takes care of instantiating it from the environment
+    """
+    def __init__(self, **kwargs):
+        self._client = _infer_auth_chain(**kwargs)
+
+# global default client
+DefaultClient: Optional[Client] = None
+
+try:
+    DefaultClient = Client()
+except NoAuthenticationFound:
+    # must call init() explicitly
+    DefaultClient = None
+
+
+def init(**kwargs):
+    global DefaultClient
+    DefaultClient = Client(**kwargs)
+
+```
+
+### model driven interface
+
+
+
+```python
+class Repository:
```

Comment:
Is there a sensible unification of commits and tags into a single `ref()`/`get_ref()` method if all that's desired is obtaining a commit ID for a given ref?

Reply:
Added!